internal/controllers/nmc_reconciler.go

```diff
@@ -520,56 +524,101 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
+func (h *nmcReconcilerHelperImpl) UpdateNodeLabelsAndRecordEvents(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error {
 	node := v1.Node{}
 	if err := h.client.Get(ctx, types.NamespacedName{Name: nmc.Name}, &node); err != nil {
 		return fmt.Errorf("could not get node %s: %v", nmc.Name, err)
 	}
 
 	// get all the kernel module ready labels of the node
-	nodeModuleReadyLabels := sets.New[string]()
+	nodeModuleReadyLabels := sets.New[types.NamespacedName]()
+
 	for label := range node.GetLabels() {
-		if utils.IsKernelModuleReadyNodeLabel(label) {
-			nodeModuleReadyLabels.Insert(label)
+		if ok, namespace, name := utils.IsKernelModuleReadyNodeLabel(label); ok {
+			nodeModuleReadyLabels.Insert(types.NamespacedName{Namespace: namespace, Name: name})
 		}
 	}
 
 	// get spec labels and their config
-	specLabels := make(map[string]kmmv1beta1.ModuleConfig)
+	specLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
 	for _, module := range nmc.Spec.Modules {
-		label := utils.GetKernelModuleReadyNodeLabel(module.Namespace, module.Name)
-		specLabels[label] = module.Config
+		specLabels[types.NamespacedName{Namespace: module.Namespace, Name: module.Name}] = module.Config
 	}
 
 	// get status labels and their config
-	statusLabels := make(map[string]kmmv1beta1.ModuleConfig)
+	statusLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
 	for _, module := range nmc.Status.Modules {
-		label := utils.GetKernelModuleReadyNodeLabel(module.Namespace, module.Name)
+		label := types.NamespacedName{Namespace: module.Namespace, Name: module.Name}
 		statusLabels[label] = kmmv1beta1.ModuleConfig{}
 		if module.Config != nil {
 			statusLabels[label] = *module.Config
 		}
 	}
 
-	patchFrom := client.MergeFrom(node.DeepCopy())
+	unloaded := make([]types.NamespacedName, 0, len(nodeModuleReadyLabels))
+	loaded := make([]types.NamespacedName, 0, len(specLabels))
+
 	// label in node but not in spec or status - should be removed
-	for label := range nodeModuleReadyLabels {
-		_, inSpec := specLabels[label]
-		_, inStatus := statusLabels[label]
+	for nsn := range nodeModuleReadyLabels {
+		_, inSpec := specLabels[nsn]
+		_, inStatus := statusLabels[nsn]
 		if !inSpec && !inStatus {
-			meta.RemoveLabel(&node, label)
+			unloaded = append(unloaded, nsn)
 		}
 	}
 
 	// label in spec and status and config equal - should be added
-	for label, specConfig := range specLabels {
-		statusConfig, ok := statusLabels[label]
-		if ok && reflect.DeepEqual(specConfig, statusConfig) {
-			meta.SetLabel(&node, label, "")
+	for nsn, specConfig := range specLabels {
+		statusConfig, ok := statusLabels[nsn]
+		if ok && reflect.DeepEqual(specConfig, statusConfig) && !nodeModuleReadyLabels.Has(nsn) {
+			loaded = append(loaded, nsn)
 		}
 	}
 
-	return h.client.Patch(ctx, &node, patchFrom)
+	patchFrom := client.MergeFrom(node.DeepCopy())
+
+	for _, nsn := range unloaded {
```

Comment:
I think that this loop and the next one a redundant: you can just call meta.RemoveLabel and meta.SetLabel in the loops where you update loaded and unloaded slices

Reply:
I thought it would be clearer to use the slices for both the events and the labels. I can change it of you insist.